add code quality tools + enforcement

Kilo59 opened the pull request:
## Changes
Add various static analysis and code quality tools including
* ~Pipenv - dependency and virtual env management~
* black - python formatter
* prospector - python linting suite 
* pre-commit - framework for various git pre-commit hooks for automatically checking and fixing common problems. The list of pre-commit hooks being applied can found by referencing the[ `.pre-commit-config.yaml` ](https://github.com/CamDavidsonPilon/lifelines/pull/561/files#diff-82c3b36709926de3a1b27bea0d8258cc) file.
* dozens of linting error fixes
* allow for failures on Python 3.4 build  


Despite the large number of files changed, there should be no functional differences between the previous version of lifelines and these changes.

-----------------------------------------------------------

## Black formatter 

The most consequential change as part of this PR is likey the addition of the [black formatter](https://github.com/ambv/black).
One of the best things about the `black` formatter is actually its lack of configuration, the one and only decision that needs to be made is what the standard line length should be. The default is `88` (what I used for this PR).
I would recommend using the default `88` line length or somewhere between the `100` and `120`.
Apart from that, you will find that the adoption of black by itself will go a long way towards reaching pep8 compliance. 

### Trying out different line-lengths

* cd to project directory
* activate virtual env
* install black
* call black on the current directory and pass the `-l` or `--line-length` arg with the desired length. 
```
pip install black
black . -l 100
```

### Excluding sections from automatic formatting

>https://github.com/ambv/black#the-black-code-style
>Black doesn't reformat blocks that start with `# fmt: off` and end with `# fmt: on`. # fmt: on/off have to be on the same level of indentation

## Pre-commit

A full list of available pre-commit hooks is available here.
https://pre-commit.com/hooks.html

### WIP

## TODO: I will add some additional explanation of the various changes in this PR soon.

CamDavidsonPilon commented on .prospector.yaml:
```diff
@@ -0,0 +1,47 @@
+inherits:
+  - strictness_medium
```
what does this refer to?

Kilo59 replied:
This just sets the baseline level of linting messages to suppress. 
If it was set to low I likely wouldn't need to turn off as many specific errors checks but it would also have turned off things we want to check for.
medium is the default, but I just put it here for visibility. 
https://prospector.landscape.io/en/master/profiles.html#builtin-profiles

CamDavidsonPilon commented on docs/conf.py:
```diff
@@ -17,224 +17,226 @@
 # If extensions (or modules to document with autodoc) are in another directory,
```
I'm guess all these python changes are black's changes?

Kilo59 replied:
I'd guess about 90% of the changes to the python files are the formatter.
The last 10% is me fixing minor linting issues that I could be sure wouldn't change the function of the python. Or adding error suppression for lint-errors that I didn't feel comfortable changing on my own.

For example, I removed a lot of unnecessary `else` statements after `return` statements.

CamDavidsonPilon commented:
What's linting the rst files? I couldn't find the source. 

> allow for failures on Python 3.4 build

Any reason for specifically choosing 3.4?

Kilo59 commented on dev-requirements.txt:
```diff
@@ -1,7 +1,56 @@
\ No newline at end of file
+-i https://pypi.org/simple
+-e .
+appdirs==1.4.3
```
A lot of these are dependencies of libs you do use, they are just being exposed here, where as before they where hidden.

However, I'm going to remove this and the `pipenv` stuff from this PR and make it a part of the automatic deployment PR. I think I'm going to use a different method for handling dependencies as `Pipenv` works better when you are only targeting 1 version of Python (which you are not).

Kilo59 replied:
If you are curious here is a full dependency graph created by pipenv of lifelines. I've removed the packages I added.

```
coveralls==1.5.1
  - coverage [required: >=3.6, installed: 4.5.2]
  - docopt [required: >=0.6.1, installed: 0.6.2]
  - requests [required: >=1.0.0, installed: 2.20.1]
    - certifi [required: >=2017.4.17, installed: 2018.11.29]
    - chardet [required: >=3.0.2,<3.1.0, installed: 3.0.4]
    - idna [required: >=2.5,<2.8, installed: 2.7]
    - urllib3 [required: >=1.21.1,<1.25, installed: 1.24.1]
lifelines==0.15.1
  - matplotlib [required: >=2.0, installed: 3.0.2]
    - cycler [required: >=0.10, installed: 0.10.0]
      - six [required: Any, installed: 1.11.0]
    - kiwisolver [required: >=1.0.1, installed: 1.0.1]
      - setuptools [required: Any, installed: 40.6.2]
    - numpy [required: >=1.10.0, installed: 1.15.4]
    - pyparsing [required: >=2.0.1,!=2.1.6,!=2.1.2,!=2.0.4, installed: 2.3.0]
    - python-dateutil [required: >=2.1, installed: 2.7.5]
      - six [required: >=1.5, installed: 1.11.0]
  - numpy [required: Any, installed: 1.15.4]
  - pandas [required: >=0.18, installed: 0.23.4]
    - numpy [required: >=1.9.0, installed: 1.15.4]
    - python-dateutil [required: >=2.5.0, installed: 2.7.5]
      - six [required: >=1.5, installed: 1.11.0]
    - pytz [required: >=2011k, installed: 2018.7]
  - scipy [required: >=1.0, installed: 1.1.0]
    - numpy [required: >=1.8.2, installed: 1.15.4]
pypandoc==1.4
  - pip [required: >=8.1.0, installed: 18.1]
  - setuptools [required: Any, installed: 40.6.2]
  - wheel [required: >=0.25.0, installed: 0.32.3]
pytest-cov==2.6.0
  - coverage [required: >=4.4, installed: 4.5.2]
  - pytest [required: >=2.9, installed: 4.0.1]
    - atomicwrites [required: >=1.0, installed: 1.2.1]
    - attrs [required: >=17.4.0, installed: 18.2.0]
    - more-itertools [required: >=4.0.0, installed: 4.3.0]
      - six [required: >=1.0.0,<2.0.0, installed: 1.11.0]
    - pluggy [required: >=0.7, installed: 0.8.0]
    - py [required: >=1.5.0, installed: 1.7.0]
    - setuptools [required: Any, installed: 40.6.2]
    - six [required: >=1.10.0, installed: 1.11.0]
seaborn==0.9.0
  - matplotlib [required: >=1.4.3, installed: 3.0.2]
    - cycler [required: >=0.10, installed: 0.10.0]
      - six [required: Any, installed: 1.11.0]
    - kiwisolver [required: >=1.0.1, installed: 1.0.1]
      - setuptools [required: Any, installed: 40.6.2]
    - numpy [required: >=1.10.0, installed: 1.15.4]
    - pyparsing [required: >=2.0.1,!=2.1.6,!=2.1.2,!=2.0.4, installed: 2.3.0]
    - python-dateutil [required: >=2.1, installed: 2.7.5]
      - six [required: >=1.5, installed: 1.11.0]
  - numpy [required: >=1.9.3, installed: 1.15.4]
  - pandas [required: >=0.15.2, installed: 0.23.4]
    - numpy [required: >=1.9.0, installed: 1.15.4]
    - python-dateutil [required: >=2.5.0, installed: 2.7.5]
      - six [required: >=1.5, installed: 1.11.0]
    - pytz [required: >=2011k, installed: 2018.7]
  - scipy [required: >=0.14.0, installed: 1.1.0]
    - numpy [required: >=1.8.2, installed: 1.15.4]
```

Kilo59 replied:
If you're curious here's 3 graphs generated by pipenv that cam help make sense of lifelines base and dev dependencies (+ everything installed in the travis build stage. 
It shows the top-level dependcies on top, so if you wanted to you should be able to pip install only those top level packages to get everything required for `lifelines`.

## Base

```
matplotlib==3.0.2
  - cycler [required: >=0.10, installed: 0.10.0]
    - six [required: Any, installed: 1.11.0]
  - kiwisolver [required: >=1.0.1, installed: 1.0.1]
    - setuptools [required: Any, installed: 40.6.2]
  - numpy [required: >=1.10.0, installed: 1.15.4]
  - pyparsing [required: >=2.0.1,!=2.1.6,!=2.1.2,!=2.0.4, installed: 2.3.0]
  - python-dateutil [required: >=2.1, installed: 2.7.5]
    - six [required: >=1.5, installed: 1.11.0]
pandas==0.23.4
  - numpy [required: >=1.9.0, installed: 1.15.4]
  - python-dateutil [required: >=2.5.0, installed: 2.7.5]
    - six [required: >=1.5, installed: 1.11.0]
  - pytz [required: >=2011k, installed: 2018.7]
scipy==1.1.0
  - numpy [required: >=1.8.2, installed: 1.15.4]
```

## Everything

```
aspy.yaml==1.1.1
  - pyyaml [required: Any, installed: ?]
coveralls==1.5.1
  - coverage [required: >=3.6, installed: 4.5.2]
  - docopt [required: >=0.6.1, installed: 0.6.2]
  - requests [required: >=1.0.0, installed: 2.20.1]
    - certifi [required: >=2017.4.17, installed: 2018.11.29]
    - chardet [required: >=3.0.2,<3.1.0, installed: 3.0.4]
    - idna [required: >=2.5,<2.8, installed: 2.7]
    - urllib3 [required: >=1.21.1,<1.25, installed: 1.24.1]
flake8==3.6.0
  - mccabe [required: >=0.6.0,<0.7.0, installed: 0.6.1]
  - pycodestyle [required: >=2.4.0,<2.5.0, installed: 2.4.0]
  - pyflakes [required: >=2.0.0,<2.1.0, installed: 2.0.0]
  - setuptools [required: >=30, installed: 40.6.2]
pypandoc==1.4
  - pip [required: >=8.1.0, installed: 18.1]
  - setuptools [required: Any, installed: 40.6.2]
  - wheel [required: >=0.25.0, installed: 0.32.3]
pytest-cov==2.6.0
  - coverage [required: >=4.4, installed: 4.5.2]
  - pytest [required: >=2.9, installed: 4.0.1]
    - atomicwrites [required: >=1.0, installed: 1.2.1]
    - attrs [required: >=17.4.0, installed: 18.2.0]
    - more-itertools [required: >=4.0.0, installed: 4.3.0]
      - six [required: >=1.0.0,<2.0.0, installed: 1.11.0]
    - pluggy [required: >=0.7, installed: 0.8.0]
    - py [required: >=1.5.0, installed: 1.7.0]
    - setuptools [required: Any, installed: 40.6.2]
    - six [required: >=1.10.0, installed: 1.11.0]
seaborn==0.9.0
  - matplotlib [required: >=1.4.3, installed: 3.0.2]
    - cycler [required: >=0.10, installed: 0.10.0]
      - six [required: Any, installed: 1.11.0]
    - kiwisolver [required: >=1.0.1, installed: 1.0.1]
      - setuptools [required: Any, installed: 40.6.2]
    - numpy [required: >=1.10.0, installed: 1.15.4]
    - pyparsing [required: >=2.0.1,!=2.1.6,!=2.1.2,!=2.0.4, installed: 2.3.0]
    - python-dateutil [required: >=2.1, installed: 2.7.5]
      - six [required: >=1.5, installed: 1.11.0]
  - numpy [required: >=1.9.3, installed: 1.15.4]
  - pandas [required: >=0.15.2, installed: 0.23.4]
    - numpy [required: >=1.9.0, installed: 1.15.4]
    - python-dateutil [required: >=2.5.0, installed: 2.7.5]
      - six [required: >=1.5, installed: 1.11.0]
    - pytz [required: >=2011k, installed: 2018.7]
  - scipy [required: >=0.14.0, installed: 1.1.0]
    - numpy [required: >=1.8.2, installed: 1.15.4]
```

## Finally

The same thing as "everything" but reversed .

```
atomicwrites==1.2.1
  - pytest==4.0.1 [requires: atomicwrites>=1.0]
    - pytest-cov==2.6.0 [requires: pytest>=2.9]
attrs==18.2.0
  - pytest==4.0.1 [requires: attrs>=17.4.0]
    - pytest-cov==2.6.0 [requires: pytest>=2.9]
certifi==2018.11.29
  - requests==2.20.1 [requires: certifi>=2017.4.17]
    - coveralls==1.5.1 [requires: requests>=1.0.0]
chardet==3.0.4
  - requests==2.20.1 [requires: chardet>=3.0.2,<3.1.0]
    - coveralls==1.5.1 [requires: requests>=1.0.0]
coverage==4.5.2
  - coveralls==1.5.1 [requires: coverage>=3.6]
  - pytest-cov==2.6.0 [requires: coverage>=4.4]
docopt==0.6.2
  - coveralls==1.5.1 [requires: docopt>=0.6.1]
idna==2.7
  - requests==2.20.1 [requires: idna>=2.5,<2.8]
    - coveralls==1.5.1 [requires: requests>=1.0.0]
mccabe==0.6.1
  - flake8==3.6.0 [requires: mccabe>=0.6.0,<0.7.0]
numpy==1.15.4
  - matplotlib==3.0.2 [requires: numpy>=1.10.0]
    - seaborn==0.9.0 [requires: matplotlib>=1.4.3]
  - pandas==0.23.4 [requires: numpy>=1.9.0]
    - seaborn==0.9.0 [requires: pandas>=0.15.2]
  - scipy==1.1.0 [requires: numpy>=1.8.2]
    - seaborn==0.9.0 [requires: scipy>=0.14.0]
  - seaborn==0.9.0 [requires: numpy>=1.9.3]
pip==18.1
  - pypandoc==1.4 [requires: pip>=8.1.0]
pluggy==0.8.0
  - pytest==4.0.1 [requires: pluggy>=0.7]
    - pytest-cov==2.6.0 [requires: pytest>=2.9]
py==1.7.0
  - pytest==4.0.1 [requires: py>=1.5.0]
    - pytest-cov==2.6.0 [requires: pytest>=2.9]
pycodestyle==2.4.0
  - flake8==3.6.0 [requires: pycodestyle>=2.4.0,<2.5.0]
pyflakes==2.0.0
  - flake8==3.6.0 [requires: pyflakes>=2.0.0,<2.1.0]
pyparsing==2.3.0
  - matplotlib==3.0.2 [requires: pyparsing>=2.0.1,!=2.1.6,!=2.1.2,!=2.0.4]
    - seaborn==0.9.0 [requires: matplotlib>=1.4.3]
pytz==2018.7
  - pandas==0.23.4 [requires: pytz>=2011k]
    - seaborn==0.9.0 [requires: pandas>=0.15.2]
pyyaml==?
  - aspy.yaml==1.1.1 [requires: pyyaml]
setuptools==40.6.2
  - flake8==3.6.0 [requires: setuptools>=30]
  - kiwisolver==1.0.1 [requires: setuptools]
    - matplotlib==3.0.2 [requires: kiwisolver>=1.0.1]
      - seaborn==0.9.0 [requires: matplotlib>=1.4.3]
  - pypandoc==1.4 [requires: setuptools]
  - pytest==4.0.1 [requires: setuptools]
    - pytest-cov==2.6.0 [requires: pytest>=2.9]
six==1.11.0
  - cycler==0.10.0 [requires: six]
    - matplotlib==3.0.2 [requires: cycler>=0.10]
      - seaborn==0.9.0 [requires: matplotlib>=1.4.3]
  - more-itertools==4.3.0 [requires: six>=1.0.0,<2.0.0]
    - pytest==4.0.1 [requires: more-itertools>=4.0.0]
      - pytest-cov==2.6.0 [requires: pytest>=2.9]
  - pytest==4.0.1 [requires: six>=1.10.0]
    - pytest-cov==2.6.0 [requires: pytest>=2.9]
  - python-dateutil==2.7.5 [requires: six>=1.5]
    - matplotlib==3.0.2 [requires: python-dateutil>=2.1]
      - seaborn==0.9.0 [requires: matplotlib>=1.4.3]
    - pandas==0.23.4 [requires: python-dateutil>=2.5.0]
      - seaborn==0.9.0 [requires: pandas>=0.15.2]
urllib3==1.24.1
  - requests==2.20.1 [requires: urllib3>=1.21.1,<1.25]
    - coveralls==1.5.1 [requires: requests>=1.0.0]
wheel==0.32.3
  - pypandoc==1.4 [requires: wheel>=0.25.0]
```

Kilo59 commented on Makefile:
```diff
@@ -1,8 +1,27 @@
+	pipenv install --dev --skip-lock
+
+test:
+	pipenv run py.test -rf -s --cov=lifelines -vv --block=False --cov-report term-missing
```
It will run the command inside a virtual env.
But I'm going to remove it.

Kilo59 commented on .prospector.yaml:
```diff
@@ -0,0 +1,47 @@
+  disable:
+    - line-too-long
+    - protected-access
+    - unused-import
```
I removed them because there were a lot of error messages raised by them. 
I meant to turn them back off or add a comment suggesting they be turned off and fixed but I forgot.

These are the lint errors returned with them off.

```
Messages
========

lifelines/estimation.py
  Line: 2
    pylint: unused-import / Unused BaseFitter imported from lifelines.fitters
  Line: 3
    pylint: unused-import / Unused WeibullFitter imported from lifelines.fitters.weibull_fitter
  Line: 4
    pylint: unused-import / Unused ExponentialFitter imported from lifelines.fitters.exponential_fitter
  Line: 5
    pylint: unused-import / Unused NelsonAalenFitter imported from lifelines.fitters.nelson_aalen_fitter
  Line: 6
    pylint: unused-import / Unused KaplanMeierFitter imported from lifelines.fitters.kaplan_meier_fitter
  Line: 7
    pylint: unused-import / Unused BreslowFlemingHarringtonFitter imported from lifelines.fitters.breslow_fleming_harrington_fitter
  Line: 10
    pylint: unused-import / Unused CoxPHFitter imported from lifelines.fitters.coxph_fitter
  Line: 11
    pylint: unused-import / Unused CoxTimeVaryingFitter imported from lifelines.fitters.cox_time_varying_fitter
  Line: 12
    pylint: unused-import / Unused AalenAdditiveFitter imported from lifelines.fitters.aalen_additive_fitter
  Line: 13
    pylint: unused-import / Unused AalenJohansenFitter imported from lifelines.fitters.aalen_johansen_fitter

lifelines/fitters/__init__.py
  Line: 20
    pylint: unused-variable / Unused variable 'estimate' (col 12)

lifelines/fitters/aalen_additive_fitter.py
  Line: 176
    pylint: unused-variable / Unused variable 'n' (col 8)
  Line: 300
    pylint: unused-variable / Unused variable 'n' (col 8)
  Line: 435
    pylint: unused-variable / Unused variable 'd' (col 11)

lifelines/fitters/aalen_johansen_fitter.py
  Line: 210
    pylint: unused-variable / Unused variable 'i' (col 12)

lifelines/fitters/cox_time_varying_fitter.py
  Line: 14
    pylint: unused-import / Unused solve imported from numpy.linalg
  Line: 19
    pylint: unused-import / Unused check_for_overlapping_intervals imported from lifelines.utils
  Line: 188
    pylint: unused-argument / Unused argument 'stop_times_events' (col 18)
  Line: 210
    pylint: unused-variable / Unused variable 'ei' (col 12)
  Line: 326
    pylint: unused-variable / Unused variable 'n' (col 8)

lifelines/fitters/coxph_fitter.py
  Line: 258
    pylint: unused-variable / Unused variable 'n' (col 8)
  Line: 611
    pylint: unused-argument / Unused argument 'T' (col 17)

lifelines/fitters/exponential_fitter.py
  Line: 48
    pylint: unused-argument / Unused argument 'entry' (col 8)

lifelines/fitters/weibull_fitter.py
  Line: 94
    pylint: unused-argument / Unused argument 'entry' (col 8)

lifelines/generate_datasets.py
  Line: 82
    pylint: unused-argument / Unused argument 't' (col 14)
    pylint: unused-argument / Unused argument 'alpha' (col 17)
  Line: 127
    pylint: unused-argument / Unused argument 'constant' (col 40)
  Line: 273
    pylint: unused-argument / Unused argument 'constant' (col 21)
    pylint: unused-argument / Unused argument 'independent' (col 37)
    pylint: unused-argument / Unused argument 'n_binary' (col 52)

lifelines/statistics.py
  Line: 331
    pylint: unused-argument / Unused argument 'test_result' (col 23)

lifelines/utils/__init__.py
  Line: 606
    pylint: unused-variable / Unused variable 'd' (col 7)
  Line: 1343
    pylint: unused-argument / Unused argument 'final_stop_time' (col 71)
    pylint: unused-variable / Unused variable 'construct_new_timeline' (col 4)



Check Information
=================
         Started: 2018-12-03 17:03:01.979813
        Finished: 2018-12-03 17:03:44.314070
      Time Taken: 42.33 seconds
       Formatter: grouped
        Profiles: .prospector.yaml, strictness_medium, strictness_high, strictness_veryhigh, no_doc_warnings, no_member_warnings, no_test_warnings
      Strictness: from profile
  Libraries Used:
       Tools Run: dodgy, mccabe, pep8, profile-validator, pyflakes, pylint, pyroma
  Messages Found: 34
```

CamDavidsonPilon replied:
seeing all this information makes me so excited to fix them haha!

Kilo59 commented:
I reverted back to using traditional files, however, I moved them to a `/reqs` folder, added a `travis-requirments.txt` file which includes the things originally not in the dev or base requirments.txt file.
In addition, I added logic to the makefile to ensure that the correct dependencies are resolved during the CI.

Specifically, I'm tried to avoid a dependency conflict that was introduced with the latest release of one of the coverage tools (I'm typing from phone and don't remember which one it was).

Kilo59 commented:
> .What's linting the rst files? I couldn't find the source.
>

Nothing, but it would have likley been altered by some combination of `trim-white-space`, the end of line fixer and/or mixed line ending checker.
 
> >allow for failures on Python 3.4 build
> 
> Any reason for specifically choosing 3.4?

As a way to ease into #559 dropping support for python3.4.

CamDavidsonPilon commented on lifelines/utils/__init__.py:
```diff
@@ -1097,8 +1097,7 @@ def concordance_value(time_a, time_b, pred_a, pred_b):
 
-    for a in range(0, len(event_times)):
-        time_a = event_times[a]
+    for a, time_a in enumerate(event_times):
```
👍

CamDavidsonPilon commented:
This is looking great! Really excited. Let me know when/if you are fully finished and wanting to merge.

Kilo59 commented:
I just changed the formatting from the default 88 to 100.
I would say the only thing left to do is for you to decide which you like better.
All thing being equal I would say it's better to go with the black default (`88`) just because that's one less thing to remember when using it.

I've also added an addional make command `pre` which will run ALL the pre-commit hooks on all the files in the repo.

CamDavidsonPilon commented:
My opinion on line lengths: let the code run free as it wants to the right :)

I'm more in favour of an unlimited line length and optimize for what's most readable. So 100 is fine for now. 


So is it black that automagically will reformat the code's line lengths? If so, that's really cool and impressive.

Kilo59 commented:
Sounds good to me. Unless you want me to change something I'm done, you can merge.

CamDavidsonPilon commented:
Great! Thank you so much for this effort. I hope you learned _something_ along the way. I’m busy the next two days (moving), but will return to this Friday or the weekend.
